File: RealtimeTranslator/Realtime/SourceBoundaryTracker.swift
import Foundation

struct SourceBoundaryTracker: Sendable {
    private(set) var candidateOffset: Int?
    private var observedSegmentGeneration: Int?

    mutating func reset() {
        candidateOffset = nil
        observedSegmentGeneration = nil
    }

    mutating func observe(
        segmentSource: String,
        deltaStart: Int,
        segmentGeneration: Int,
        pair: LanguagePair,
        currentLanguage: SpokenLanguage,
        reverseEvidenceCount: Int
    ) {
        if observedSegmentGeneration != segmentGeneration {
            reset()
            observedSegmentGeneration = segmentGeneration
        }

        let sourceLength = segmentSource.utf16.count
        let start = min(max(deltaStart, 0), sourceLength)
        if pair == .enEs {
            observeEnEs(
                segmentSource: segmentSource,
                deltaStart: start,
                currentLanguage: currentLanguage,
                reverseEvidenceCount: reverseEvidenceCount
            )
        } else {
            observeScriptPair(
                segmentSource: segmentSource,
                deltaStart: start,
                currentLanguage: currentLanguage
            )
        }
    }

    private mutating func observeScriptPair(
        segmentSource: String,
        deltaStart: Int,
        currentLanguage: SpokenLanguage
    ) {
        let oppositeIsJapanese = currentLanguage != .japanese
        let oppositeIsLatin = currentLanguage == .japanese
        let entries = scalarEntries(in: segmentSource)

        for entry in entries where entry.offset >= deltaStart {
            let isJapanese = Self.isJapanese(entry.scalar)
            let isLatin = SpokenLanguageDetector.isLatinWordScalar(entry.scalar)
            let isOpposite = oppositeIsJapanese ? isJapanese : isLatin
            let isOwn = oppositeIsJapanese ? isLatin : isJapanese

            if candidateOffset == nil, isOpposite {
                candidateOffset = moveBackwardOverNewSidePrefix(
                    candidateOffset: entry.offset,
                    entries: entries
                )
            } else if candidateOffset != nil, isOwn {
                candidateOffset = nil
            }
        }
    }

    private mutating func observeEnEs(
        segmentSource: String,
        deltaStart _: Int,
        currentLanguage: SpokenLanguage,
        reverseEvidenceCount: Int
    ) {
        guard reverseEvidenceCount > 0 else {
            candidateOffset = nil
            return
        }
        guard reverseEvidenceCount == 1 else { return }

        let reverseLanguage = currentLanguage == .english ? SpokenLanguage.spanish : .english
        if let candidateOffset,
           firstCueStarting(atOrAfter: candidateOffset, in: segmentSource) == reverseLanguage
        {
            return
        }

        let spans = SpokenLanguageDetector.wordSpans(in: segmentSource)
        let recentStart = SpokenLanguageDetector.recentWordWindowStart(in: segmentSource)
        let recentStartOffset = recentStart.utf16Offset(in: segmentSource)
        let recentEnd = segmentSource.utf16.count
        let recentSpans = spans.filter {
            $0.lowerBound >= recentStartOffset && $0.upperBound <= recentEnd
        }
        let firstReverseWord = recentSpans.first {
            cueLanguage(String(segmentSource.utf16Slice($0))) == reverseLanguage
        }?.lowerBound
        let firstReverseMark = firstStandaloneSpanishMark(
            in: segmentSource,
            from: recentStartOffset,
            to: recentEnd
        )
        let cueStart = [firstReverseWord, firstReverseMark]
            .compactMap { $0 }
            .min()

        guard let cueStart else {
            candidateOffset = nil
            return
        }

        let sentenceStart = sentenceStart(
            in: segmentSource,
            windowStart: recentStartOffset,
            before: cueStart
        )
        let hasCurrentCue = recentSpans.contains {
            $0.lowerBound >= sentenceStart
                && $0.lowerBound < cueStart
                && cueLanguage(String(segmentSource.utf16Slice($0)))
                    == currentLanguage
        }
        let rawCandidate = hasCurrentCue ? cueStart : sentenceStart
        candidateOffset = moveBackwardOverNewSidePrefix(
            candidateOffset: rawCandidate,
            entries: scalarEntries(in: segmentSource)
        )
    }

    private func firstCueStarting(
        atOrAfter offset: Int,
        in source: String
    ) -> SpokenLanguage? {
        let firstWord = SpokenLanguageDetector.wordSpans(in: source)
            .filter { $0.lowerBound >= offset }
            .compactMap { span -> (Int, SpokenLanguage)? in
                guard let language = cueLanguage(String(source.utf16Slice(span))) else {
                    return nil
                }
                return (span.lowerBound, language)
            }
            .first
        let firstMark = firstStandaloneSpanishMark(
            in: source,
            from: offset,
            to: source.utf16.count
        ).map { ($0, SpokenLanguage.spanish) }
        switch [firstWord, firstMark].compactMap({ $0 }).min(by: { $0.0 < $1.0 })?.1 {
        case .some(let language):
            return language
        case .none:
            return nil
        }
    }

    private func cueLanguage(_ word: String) -> SpokenLanguage? {
        let lower = word.lowercased(with: Locale(identifier: "en_US_POSIX"))
        if SpokenLanguageDetector.englishExclusiveWords.contains(lower) {
            return .english
        }
        if SpokenLanguageDetector.spanishExclusiveWords.contains(lower)
            || word.unicodeScalars.contains(where: Self.isSpanishAccentOrN)
        {
            return .spanish
        }
        return nil
    }

    private func sentenceStart(in source: String, windowStart: Int, before offset: Int) -> Int {
        var result = windowStart
        for entry in scalarEntries(in: source)
            where entry.offset >= windowStart && entry.offset < offset
        {
            if Self.isSentenceTerminator(entry.scalar) {
                result = entry.offset + entry.scalar.utf16.count
            }
        }
        return result
    }

    private func firstStandaloneSpanishMark(
        in source: String,
        from start: Int,
        to end: Int
    ) -> Int? {
        scalarEntries(in: source).first {
            $0.offset >= start
                && $0.offset < end
                && ($0.scalar.value == 0x00BF || $0.scalar.value == 0x00A1)
        }?.offset
    }

    private func moveBackwardOverNewSidePrefix(
        candidateOffset: Int,
        entries: [(offset: Int, scalar: Unicode.Scalar)]
    ) -> Int {
        var result = candidateOffset
        var index = entries.firstIndex { $0.offset >= candidateOffset } ?? entries.count
        while index > 0 {
            let previous = entries[index - 1]
            guard CharacterSet.whitespacesAndNewlines.contains(previous.scalar)
                || previous.scalar.value == 0x00BF
                || previous.scalar.value == 0x00A1
            else {
                break
            }
            result = previous.offset
            index -= 1
        }
        return result
    }

    private func scalarEntries(in source: String) -> [(offset: Int, scalar: Unicode.Scalar)] {
        var result: [(offset: Int, scalar: Unicode.Scalar)] = []
        var offset = 0
        for scalar in source.unicodeScalars {
            result.append((offset: offset, scalar: scalar))
            offset += scalar.utf16.count
        }
        return result
    }

    private static func isJapanese(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.value {
        case 0x3040...0x30FF, 0x3400...0x4DBF, 0x4E00...0x9FFF:
            return true
        default:
            return false
        }
    }

    private static func isSpanishAccentOrN(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.value {
        case 0x00E1, 0x00E9, 0x00ED, 0x00F3, 0x00FA, 0x00FC,
             0x00C1, 0x00C9, 0x00CD, 0x00D3, 0x00DA, 0x00DC,
             0x00F1, 0x00D1:
            return true
        default:
            return false
        }
    }

    private static func isSentenceTerminator(_ scalar: Unicode.Scalar) -> Bool {
        scalar.value == 0x002E || scalar.value == 0x0021 || scalar.value == 0x003F
            || scalar.value == 0x3002 || scalar.value == 0xFF01 || scalar.value == 0xFF1F
    }
}

private extension String {
    func utf16Slice(_ range: Range<Int>) -> String {
        let start = String.Index(utf16Offset: range.lowerBound, in: self)
        let end = String.Index(utf16Offset: range.upperBound, in: self)
        return String(self[start..<end])
    }
}

File: windows/src/RealtimeTranslator.Core/Realtime/SourceBoundaryTracker.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.Unicode;
using RealtimeTranslator.Core.Audio;

namespace RealtimeTranslator.Core.Realtime;

public sealed class SourceBoundaryTracker
{
    public int? CandidateOffset { get; private set; }

    private int? _observedSegmentGeneration;

    public void Reset()
    {
        CandidateOffset = null;
        _observedSegmentGeneration = null;
    }

    public void Observe(
        string segmentSource,
        int deltaStart,
        int segmentGeneration,
        LanguagePair pair,
        SpokenLanguage currentLanguage,
        int reverseEvidenceCount)
    {
        ArgumentNullException.ThrowIfNull(segmentSource);
        if (_observedSegmentGeneration != segmentGeneration)
        {
            Reset();
            _observedSegmentGeneration = segmentGeneration;
        }

        var start = Math.Clamp(deltaStart, 0, segmentSource.Length);
        if (pair == LanguagePair.EnEs)
        {
            ObserveEnEs(segmentSource, currentLanguage, reverseEvidenceCount);
        }
        else
        {
            ObserveScriptPair(segmentSource, start, currentLanguage);
        }
    }

    private void ObserveScriptPair(
        string source,
        int deltaStart,
        SpokenLanguage currentLanguage)
    {
        var oppositeIsJapanese = currentLanguage != SpokenLanguage.Japanese;
        var entries = ScalarEntries(source);
        foreach (var entry in entries.Where(entry => entry.Offset >= deltaStart))
        {
            var isJapanese = IsJapanese(entry.Rune);
            var isLatin = SpokenLanguageDetector.IsLatinWordScalar(entry.Rune);
            var isOpposite = oppositeIsJapanese ? isJapanese : isLatin;
            var isOwn = oppositeIsJapanese ? isLatin : isJapanese;

            if (CandidateOffset is null && isOpposite)
            {
                CandidateOffset = MoveBackwardOverNewSidePrefix(
                    entry.Offset,
                    entries);
            }
            else if (CandidateOffset is not null && isOwn)
            {
                CandidateOffset = null;
            }
        }
    }

    private void ObserveEnEs(
        string source,
        SpokenLanguage currentLanguage,
        int reverseEvidenceCount)
    {
        if (reverseEvidenceCount == 0)
        {
            CandidateOffset = null;
            return;
        }

        if (reverseEvidenceCount != 1)
        {
            return;
        }

        var reverseLanguage = currentLanguage == SpokenLanguage.English
            ? SpokenLanguage.Spanish
            : SpokenLanguage.English;
        var entries = ScalarEntries(source);
        if (CandidateOffset is { } candidate
            && FirstCueStartingAtOrAfter(source, candidate, entries) == reverseLanguage)
        {
            return;
        }

        var spans = SpokenLanguageDetector.WordSpans(source);
        var windowStart = SpokenLanguageDetector.RecentWordWindowStart(source);
        var recentSpans = spans
            .Where(span => span.Start >= windowStart)
            .ToArray();
        var firstReverseWord = recentSpans
            .Select(span => (Span: span, Language: CueLanguage(source[span.Start..span.End])))
            .Where(value => value.Language == reverseLanguage)
            .Select(value => (int?)value.Span.Start)
            .FirstOrDefault();
        var firstReverseMark = FirstStandaloneSpanishMark(
            source,
            windowStart,
            source.Length,
            entries);
        var cueStart = new[] { firstReverseWord, firstReverseMark }
            .Where(value => value is not null)
            .Select(value => value!.Value)
            .DefaultIfEmpty()
            .Min();
        var hasCue = firstReverseWord is not null || firstReverseMark is not null;

        if (!hasCue)
        {
            CandidateOffset = null;
            return;
        }

        var sentenceStart = SentenceStart(source, windowStart, cueStart, entries);
        var hasCurrentCue = recentSpans.Any(span =>
            span.Start >= sentenceStart
            && span.Start < cueStart
            && CueLanguage(source[span.Start..span.End]) == currentLanguage);
        var rawCandidate = hasCurrentCue ? cueStart : sentenceStart;
        CandidateOffset = MoveBackwardOverNewSidePrefix(rawCandidate, entries);
    }

    private static SpokenLanguage? FirstCueStartingAtOrAfter(
        string source,
        int offset,
        List<(int Offset, Rune Rune)> entries)
    {
        int? firstWordOffset = null;
        SpokenLanguage? firstWordLanguage = null;
        foreach (var span in SpokenLanguageDetector.WordSpans(source)
            .Where(span => span.Start >= offset))
        {
            var language = CueLanguage(source[span.Start..span.End]);
            if (language is not null)
            {
                firstWordOffset = span.Start;
                firstWordLanguage = language;
                break;
            }
        }
        var firstMark = FirstStandaloneSpanishMark(source, offset, source.Length, entries);
        if (firstWordLanguage is null && firstMark is null)
        {
            return null;
        }

        if (firstMark is null || (firstWordOffset is not null && firstWordOffset < firstMark))
        {
            return firstWordLanguage;
        }

        return SpokenLanguage.Spanish;
    }

    private static SpokenLanguage? CueLanguage(string word)
    {
        var lower = word.ToLowerInvariant();
        if (SpokenLanguageDetector.EnglishExclusiveWords.Contains(lower))
        {
            return SpokenLanguage.English;
        }

        if (SpokenLanguageDetector.SpanishExclusiveWords.Contains(lower)
            || word.EnumerateRunes().Any(IsSpanishAccentOrN))
        {
            return SpokenLanguage.Spanish;
        }

        return null;
    }

    private static int SentenceStart(
        string source,
        int windowStart,
        int before,
        List<(int Offset, Rune Rune)> entries)
    {
        var result = windowStart;
        foreach (var entry in entries
            .Where(entry => entry.Offset >= windowStart && entry.Offset < before))
        {
            if (IsSentenceTerminator(entry.Rune))
            {
                result = entry.Offset + entry.Rune.Utf16SequenceLength;
            }
        }

        return result;
    }

    private static int? FirstStandaloneSpanishMark(
        string source,
        int start,
        int end,
        List<(int Offset, Rune Rune)> entries)
    {
        foreach (var entry in entries)
        {
            if (entry.Offset >= start
                && entry.Offset < end
                && (entry.Rune.Value == 0x00BF || entry.Rune.Value == 0x00A1))
            {
                return entry.Offset;
            }
        }

        return null;
    }

    private static int MoveBackwardOverNewSidePrefix(
        int candidate,
        List<(int Offset, Rune Rune)> entries)
    {
        var result = candidate;
        var index = 0;
        while (index < entries.Count && entries[index].Offset < candidate)
        {
            index++;
        }

        while (index > 0)
        {
            var previous = entries[index - 1];
            if (!Rune.IsWhiteSpace(previous.Rune)
                && previous.Rune.Value != 0x00BF
                && previous.Rune.Value != 0x00A1)
            {
                break;
            }

            result = previous.Offset;
            index--;
        }

        return result;
    }

    private static List<(int Offset, Rune Rune)> ScalarEntries(string source)
    {
        var entries = new List<(int Offset, Rune Rune)>();
        var offset = 0;
        foreach (var rune in source.EnumerateRunes())
        {
            entries.Add((offset, rune));
            offset += rune.Utf16SequenceLength;
        }

        return entries;
    }

    private static bool IsJapanese(Rune rune) =>
        rune.Value is >= 0x3040 and <= 0x30FF
            or >= 0x3400 and <= 0x4DBF
            or >= 0x4E00 and <= 0x9FFF;

    private static bool IsSpanishAccentOrN(Rune rune) =>
        rune.Value is 0x00E1 or 0x00E9 or 0x00ED or 0x00F3 or 0x00FA or 0x00FC
            or 0x00C1 or 0x00C9 or 0x00CD or 0x00D3 or 0x00DA or 0x00DC
            or 0x00F1 or 0x00D1;

    private static bool IsSentenceTerminator(Rune rune) =>
        rune.Value is 0x002E or 0x0021 or 0x003F or 0x3002 or 0xFF01 or 0xFF1F;
}
